### PR TITLE
feat(Table): Add `useGlobalFilter`

### PR DESCRIPTION
**Pull request by DziugasRam**
Added useGlobalFilter plugin hook to the table to allow table wide search using an external search bar

## Checklist

- [x] Add meaningful unit tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- [x] Add component features demo in Storybook (different stories)
- [x] Approve test images for new stories
- [] ~~Add screenshots of the key elements of the component~~

### CODE REVIEW

**Comment by gretanausedaite** [comment]
Hi! Thank you for the contribution!

Some comments:
- PR should follow title convention: feat(Table): Add `useGlobalFilter`
- Unit tests are missing
- Visual tests are missing

You can find information on how to add tests in our [contributing guide](https://github.com/iTwin/iTwinUI-react/blob/main/CONTRIBUTING.md).

If you have any questions, feel free to reach out or add them here.

File: packages/iTwinUI-react/src/core/Table/Table.tsx [hunk 1]
```diff
@@ -178,6 +179,11 @@ export type TableProps<
    * Must be memoized.
    */
   onFilter?: (filters: TableFilterValue<T>[], state: TableState<T>) => void;
+  /**
+   * Value used for global filtering.
+   * Use with `globalFilter` and/or `manualGlobalFilter` to handle filtering yourself e.g. filter in server-side.
+   */
+  globalFilterValue?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
```
Comment (gretanausedaite) [inline packages/iTwinUI-react/src/core/Table/Table.tsx:186]: Is there a reason you chose this name?
`globalFilter` might be more suitable for prop name.

File: apps/storybook/src/Table.stories.tsx [hunk 2]
```diff
@@ -570,6 +571,96 @@ Filters.args = {
...
+          },
+        ],
+      },
+    ],
+    [],
+  );
+
+  const data = useMemo<TableStoryDataType[]>(
+    () => [
+      { name: 'Name1', description: 'Description1' },
```
Comment (gretanausedaite) [inline apps/storybook/src/Table.stories.tsx:619]: Could you change `Name` vs `Description` numbers, eg. Name2, Description5, to add more variety. This way people can see that it is filtering both columns.

File: packages/iTwinUI-react/src/core/Table/Table.tsx (hunk 1 above)
Comment (elephantcatdog) [inline packages/iTwinUI-react/src/core/Table/Table.tsx:186]: Why does this need to be `any`?

**Review by elephantcatdog — COMMENTED** [review]
Nice tests!

File: packages/iTwinUI-react/src/core/Table/Table.tsx (hunk 1 above)
Comment (DziugasRam) [reply packages/iTwinUI-react/src/core/Table/Table.tsx:186]: custom filter function can be used with any filtering value type as shown [here](https://react-table-v7.tanstack.com/docs/api/useGlobalFilter#table-options), so I didn't want to limit it to string only

File: packages/iTwinUI-react/src/core/Table/Table.tsx [hunk 3]
```diff
@@ -178,6 +179,11 @@ export type TableProps<
    * Must be memoized.
    */
   onFilter?: (filters: TableFilterValue<T>[], state: TableState<T>) => void;
+  /**
+   * Value used for global filtering.
+   * Use with `globalFilter` and/or `manualGlobalFilter` to handle filtering yourself e.g. filter in server-side.
```
Comment (gretanausedaite) [inline packages/iTwinUI-react/src/core/Table/Table.tsx:184]: Add comment that this value must be memoized.

**Review by gretanausedaite — COMMENTED** [review]
Generally looks good.
You need to uncomment global filter props in `src/types/react-table-config.ts` to expose react-table props to user.
After quick test, I found that global filter does not work on `Full2` story. It might be because it has virtualized rows or subrows.

File: packages/iTwinUI-react/src/core/Table/Table.tsx [hunk 4]
```diff
@@ -498,6 +505,7 @@ export const Table = <
     },
     useFlexLayout,
     useResizeColumns(ownerDocument),
+    useGlobalFilter,
     useFilters,
     useSubRowFiltering(hasAnySubRows),
```
Comment (gretanausedaite) [inline packages/iTwinUI-react/src/core/Table/Table.tsx:510]: globalFilter with subrows would be fixed if you moved `useGlobalFilter` after `useSubRowFiltering`

File: packages/iTwinUI-react/src/types/react-table-config.ts [hunk 5]
```diff
@@ -86,9 +86,9 @@ declare module 'react-table' {
       UseRowSelectOptions<D>,
       UseExpandedOptions<D>,
       UseFiltersOptions<D>,
-      // UseGlobalFiltersOptions<D>,
       // UseGroupByOptions<D>,
       UsePaginationOptions<D>,
+      UseGlobalFiltersOptions<D>,
```
Comment (gretanausedaite) [inline packages/iTwinUI-react/src/types/react-table-config.ts:91]: `UseGlobalFiltersColumnOptions` would also be useful for disabling global filter for individual columns.

**Review by gretanausedaite — COMMENTED** [review]
It would be nice if you added condition to show empty table content when global filter finds nothing. Table.tsx line 930 (something similar to `filteredFlatRows.length === 0 && areFiltersSet` ).
Have you checked if `disableGlobalFilter` prop is working?

**Comment by elephantcatdog** [comment]
Please update the branch when you get a chance

**Review by gretanausedaite — APPROVED** [review]
Looks good to me.

File: packages/iTwinUI-react/src/core/Table/Table.tsx [hunk 6]
```diff
@@ -927,7 +928,7 @@ export const Table = <
             </div>
           )}
           {!isLoading &&
-            (data.length === 0 || filteredFlatRows.length === 0) &&
+            (data.length === 0 || rows.length === 0) &&
```
Comment (elephantcatdog) [inline packages/iTwinUI-react/src/core/Table/Table.tsx:931]: Why did you change from `filteredFlatRows` to `rows` here?

Comment (DziugasRam) [reply packages/iTwinUI-react/src/core/Table/Table.tsx:931]: If I'm not mistaken, `filteredFlatRows` is rows after column filters are applied and `rows` are final rows after all the filters are applied

**Review by elephantcatdog — APPROVED** [review]
Looks good to me! You'll have to merge the main branch in again, ha
